File: libnd4j/include/ops/declarable/platform/llamacpp/flash_attention.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// @author Adam Gibson
//

#include <ops/declarable/OpRegistrator.h>
#include <ops/declarable/PlatformHelper.h>
#include <system/platform_boilerplate.h>

#include "llamacppUtils.h"

#if HAVE_LLAMACPP

namespace sd {
namespace ops {
namespace platforms {

/**
 * Flash Attention using GGML
 *
 * Flash Attention computes attention more efficiently by:
 * 1. Tiling the attention computation
 * 2. Avoiding materialization of the full attention matrix
 * 3. Using online softmax for numerical stability
 *
 * Inputs:
 *   0: Q (query) [batch, heads, seq_len, head_dim]
 *   1: K (key)   [batch, heads, kv_seq_len, head_dim]
 *   2: V (value) [batch, heads, kv_seq_len, head_dim]
 *   3: mask (optional) [batch, 1, seq_len, kv_seq_len]
 *
 * Output:
 *   0: attention output [batch, heads, seq_len, head_dim]
 */
static void flashAttentionLlamaCpp(NDArray* query, NDArray* key, NDArray* value,
                                    NDArray* mask, NDArray* output, float scale) {
    llamacppUtils::GgmlContextGuard ctx(128 * 1024 * 1024);  // 128MB workspace

    struct ggml_tensor* ggml_q = llamacppUtils::createGgmlTensor(ctx, query, "query");
    struct ggml_tensor* ggml_k = llamacppUtils::createGgmlTensor(ctx, key, "key");
    struct ggml_tensor* ggml_v = llamacppUtils::createGgmlTensor(ctx, value, "value");

    // Use GGML's flash attention implementation
    struct ggml_tensor* ggml_output = ggml_flash_attn_ext(
        ctx, ggml_q, ggml_k, ggml_v,
        mask != nullptr ? llamacppUtils::createGgmlTensor(ctx, mask, "mask") : nullptr,
        scale,
        0.0f,  // max_bias (for ALiBi, set to 0 for standard attention)
        0.0f   // logit_softcap (no capping)
    );

    ggml_set_name(ggml_output, "flash_attn_output");

    // Build and execute computation graph
    struct ggml_cgraph* graph = ggml_new_graph(ctx);
    ggml_build_forward_expand(graph, ggml_output);
    llamacppUtils::executeGgmlGraph(ctx, graph);

    llamacppUtils::copyGgmlToNDArray(ggml_output, output);
}

//////////////////////////////////////////////////////////////////////////
PLATFORM_IMPL(flash_attention, ENGINE_CPU) {
    auto query = INPUT_VARIABLE(0);
    auto key = INPUT_VARIABLE(1);
    auto value = INPUT_VARIABLE(2);
    NDArray* mask = block.width() > 3 ? INPUT_VARIABLE(3) : nullptr;
    auto output = OUTPUT_VARIABLE(0);

    if (query->isEmpty() || key->isEmpty() || value->isEmpty()) {
        return sd::Status::OK;
    }

    // Scale factor (default: 1/sqrt(head_dim))
    float scale = block.getTArguments()->size() > 0 ?
        T_ARG(0) :
        1.0f / std::sqrt(static_cast<float>(query->sizeAt(-1)));

    flashAttentionLlamaCpp(query, key, value, mask, output, scale);

    return sd::Status::OK;
}

//////////////////////////////////////////////////////////////////////////
PLATFORM_CHECK(flash_attention, ENGINE_CPU) {
    auto query = INPUT_VARIABLE(0);
    auto key = INPUT_VARIABLE(1);
    auto value = INPUT_VARIABLE(2);
    auto output = OUTPUT_VARIABLE(0);

    Requirements req("LLAMACPP FLASH_ATTENTION OP");

    req.expectTrue(block.isUseLLAMACPP(), IS_USE_LLAMACPP_MSG);

    // Check supported data types
    req.expectTrue(
        makeInfoVariable(
            [query, key, value, output] {
                return llamacppUtils::isSupportedType(query->dataType()) &&
                       llamacppUtils::isSupportedType(key->dataType()) &&
                       llamacppUtils::isSupportedType(value->dataType()) &&
                       llamacppUtils::isSupportedType(output->dataType());
            },
            TYPECHECK_MSG),
        NO_MSG);

    // Flash attention requires 4D tensors: [batch, heads, seq_len, head_dim]
    req.expectEq(makeInfoVariable(query->rankOf(), RANK_MSG_INPUT0), 4);
    req.expectEq(makeInfoVariable(key->rankOf(), RANK_MSG_INPUT1), 4);
    req.expectEq(makeInfoVariable(value->rankOf(), RANK_MSG_INPUT2), 4);

    req.logTheSuccess();
    return req;
}

}  // namespace platforms
}  // namespace ops
}  // namespace sd

#endif  // HAVE_LLAMACPP

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/FlashAttention.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.transforms.custom;

import lombok.NoArgsConstructor;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.common.base.Preconditions;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Flash Attention - Memory-efficient attention computation.
 *
 * Implements the Flash Attention algorithm which:
 * 1. Processes attention in tiles to reduce memory from O(N^2) to O(N)
 * 2. Uses online softmax for numerical stability
 * 3. Never materializes the full attention matrix
 *
 * Reference: "FlashAttention: Fast and Memory-Efficient Exact Attention with IO-Awareness"
 * https://arxiv.org/abs/2205.14135
 *
 * @author Adam Gibson
 */
@NoArgsConstructor
public class FlashAttention extends DynamicCustomOp {
    private boolean isCausal = true;
    private double scale = 0.0;  // 0 = auto (1/sqrt(head_dim))
    private int numHeads = 8;
    private int numKvHeads = 0;  // 0 = same as numHeads

    /**
     * Create a flash attention operation.
     *
     * @param query Query tensor [batch, seq_len, num_heads, head_dim]
     * @param key Key tensor [batch, kv_len, num_kv_heads, head_dim]
     * @param value Value tensor [batch, kv_len, num_kv_heads, head_dim]
     * @param scale Scale factor for attention scores (0 = auto: 1/sqrt(head_dim))
     * @param isCausal Whether to apply causal (lower triangular) mask
     * @param numHeads Number of query attention heads
     * @param numKvHeads Number of KV heads (0 = same as numHeads, for GQA use smaller value)
     */
    public FlashAttention(INDArray query, INDArray key, INDArray value,
                          double scale, boolean isCausal, int numHeads, int numKvHeads) {
        super(wrapFilterNull(query, key, value), null);
        this.scale = scale;
        this.isCausal = isCausal;
        this.numHeads = numHeads;
        this.numKvHeads = numKvHeads;
        addTArgument(scale);
        addBArgument(isCausal);
        addIArgument(numHeads, numKvHeads);
    }

    /**
     * Create a flash attention operation (simplified constructor).
     *
     * @param query Query tensor [batch, seq_len, num_heads, head_dim]
     * @param key Key tensor [batch, kv_len, num_kv_heads, head_dim]
     * @param value Value tensor [batch, kv_len, num_kv_heads, head_dim]
     * @param isCausal Whether to apply causal (lower triangular) mask
     * @param scale Scale factor for attention scores (0 = auto: 1/sqrt(head_dim))
     */
    public FlashAttention(INDArray query, INDArray key, INDArray value,
                          boolean isCausal, double scale) {
        super(wrapFilterNull(query, key, value), null);
        this.isCausal = isCausal;
        this.scale = scale;
        this.numHeads = (int) query.size(2);
        this.numKvHeads = (int) key.size(2);
        addTArgument(scale);
        addBArgument(isCausal);
        addIArgument(numHeads, numKvHeads);
    }

    /**
     * Create a flash attention operation with mask.
     *
     * @param query Query tensor [batch, seq_len, num_heads, head_dim]
     * @param key Key tensor [batch, kv_len, num_kv_heads, head_dim]
     * @param value Value tensor [batch, kv_len, num_kv_heads, head_dim]
     * @param mask Attention mask [batch, 1, seq_len, kv_len] (optional)
     * @param isCausal Whether to apply causal (lower triangular) mask
     * @param scale Scale factor for attention scores (0 = auto: 1/sqrt(head_dim))
     */
    public FlashAttention(INDArray query, INDArray key, INDArray value, INDArray mask,
                          boolean isCausal, double scale) {
        super(wrapFilterNull(query, key, value, mask), null);
        this.isCausal = isCausal;
        this.scale = scale;
        this.numHeads = (int) query.size(2);
        this.numKvHeads = (int) key.size(2);
        addTArgument(scale);
        addBArgument(isCausal);
        addIArgument(numHeads, numKvHeads);
    }

    /**
     * Create a flash attention operation for SameDiff.
     */
    public FlashAttention(SameDiff sd, SDVariable query, SDVariable key, SDVariable value,
                          double scale, boolean isCausal, int numHeads, int numKvHeads) {
        super(null, sd, new SDVariable[]{query, key, value}, false);
        this.scale = scale;
        this.isCausal = isCausal;
        this.numHeads = numHeads;
        this.numKvHeads = numKvHeads;
        addTArgument(scale);
        addBArgument(isCausal);
        addIArgument(numHeads, numKvHeads);
    }

    /**
     * Create a flash attention operation for SameDiff (simplified).
     */
    public FlashAttention(SameDiff sd, SDVariable query, SDVariable key, SDVariable value,
                          boolean isCausal, double scale) {
        super(null, sd, new SDVariable[]{query, key, value}, false);
        this.isCausal = isCausal;
        this.scale = scale;
        addTArgument(scale);
        addBArgument(isCausal);
        addIArgument(numHeads, numKvHeads);
    }

    /**
     * Create a flash attention operation for SameDiff with mask.
     */
    public FlashAttention(SameDiff sd, SDVariable query, SDVariable key, SDVariable value,
                          SDVariable mask, boolean isCausal, double scale) {
        super(null, sd, mask != null ?
              new SDVariable[]{query, key, value, mask} :
              new SDVariable[]{query, key, value}, false);
        this.isCausal = isCausal;
        this.scale = scale;
        addTArgument(scale);
        addBArgument(isCausal);
        addIArgument(numHeads, numKvHeads);
    }

    @Override
    public void configureFromArguments() {
        super.configureFromArguments();
        // T_ARG: scale
        if (!tArguments.isEmpty()) {
            this.scale = tArguments.get(0);
        }
        // B_ARG: isCausal
        if (!bArguments.isEmpty()) {
            this.isCausal = bArguments.get(0);
        }
        // I_ARG: numHeads, numKvHeads
        if (!iArguments.isEmpty()) {
            this.numHeads = iArguments.get(0).intValue();
        }
        if (iArguments.size() > 1) {
            this.numKvHeads = iArguments.get(1).intValue();
        }
    }

    @Override
    public String opName() {
        return "flash_attention";
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> gradient) {
        return Arrays.asList(new FlashAttentionBp(sameDiff,
                arg(0),              // query
                arg(1),              // key
                arg(2),              // value
                gradient.get(0),     // gradOutput
                isCausal,
                scale).outputVariables());
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        DataType first = dataTypes.get(0);
        Preconditions.checkState(first.isFPType(),
            "Input datatype must be a floating point type, got %s", first);
        return Collections.singletonList(first);
    }

    @Override
    public int getNumOutputs() {
        return 1;
    }
}
